kep/119-node-resource-topology-aware-scheduling/README.md

```diff
@@ -0,0 +1,335 @@
+Plugin would check for the ability to run pod only in case of single-numa-node policy on the
+node, since it is the most strict policy, it implies that the launch on the node with
+other existing policies will be successful if the condition for single-numa-node policy passed for the worker node.
+Proposed plugin will use [CRD][1] to identify which topology policy is enabled on the node.
+To work, this plugin requires topology information of the available resource on the worker nodes.
+
+## Topology format
+
+Available resources with topology of the node should be stored in CRD. Format of the topology described
+[in this document][1].
+
+[The daemon][5] which runs outside of the kubelet will collect all necessary information on running pods, based on allocatable resources of the node and consumed resources by pods it will provide available resources in CRD, where one CRD instance represents one worker node. The name of the CRD instance is the name of the worker node.
+
+## CRD API
+
+[Code][3] responsible for working with NodeResourceTopology CRD API will imported in the scheduler-plugins repo.
+
+## Plugin implementation details
+
+Since topology of the node is stored in the CRD, kube-scheduler should be subscribed for updates of appropriate CRD type. Kube-scheduler will use informers which will be generated with the name NodeTopologyInformer. NodeTopologyInformer will run in NodeResourceTopologyMatch plugin.
+
+### Topology information in the NodeResourceTopologyMatch plugin
+
+Once NodeResourceTopology is received NodeResourceTopologyMatch plugin keeps it in its own state of type
+NodeTopologyMap. This state will be used every time when scheduler needs to make a decidion based on node topology.
+
+```go
+
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// NodeResourceTopologyList is a list of NodeResourceTopology resources
+type NodeResourceTopologyList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+
+	Items []NodeResourceTopology `json:"items"`
+}
+
+// NodeResourceTopology is a specification for a NodeResourceTopology resource
+type NodeResourceTopology struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	TopologyPolicies []string `json:"topologyPolicies"`
+	Zones            ZoneList `json:"zones"`
+}
+
+// Zone is the spec for a NodeResourceTopology resource
+type Zone struct {
+	Name       string           `json:"name"`
+	Type       string           `json:"type"`
+	Parent     string           `json:"parent,omitempty"`
+	Costs      CostList         `json:"costs,omitempty"`
+	Attributes AttributeList    `json:"attributes,omitempty"`
+	Resources  ResourceInfoList `json:"resources,omitempty"`
+}
+
+type ZoneList []Zone
+
+type ResourceInfo struct {
+	Name        string             `json:"name"`
+	Allocatable intstr.IntOrString `json:"allocatable"`
```

Comment:
How can we maintain the consistency for `NodeResourceTopology.Allocatable` and reserved CPUs. Let's suppose if schedule a pod on the node. 

+ In Filter and Score extension, we use the `NodeResourceTopology.Allocatable` to filter and select the host.
+ In Reserve extension, we reserve several CPUs for the current pod (kubelet don't know the reserved CPUs)
+ In the PostBind extension, we should release the CPUs as Kubelet would update the CRD to reflect the CPU usage, e.g. reduce the allocatable field.

Since kubelet and scheduler do not talk to each other, is it possible that there could be a small timeslot between releasing the CPUs in the PostBind extension and kubelet update the CRD? And is a wrong scheduling decision could be made in that gap?

Reply:
Yes, this KEP didn't cover issue you mentioned, but it should. We can't follow approach of noderesources/fit.go built-in plugin, where plugin relies on NodeInfo.Allocatable, which updates once pod nominated, and it reduces the gap. We can't follow it, because we can't predict which NUMA node will be chosen on the worker node.
I think in future we can try to add some "hint", e.g. annotation at PostBind stage to inform kubelet which NUMA node was selected and use it to reduce amount of available resource of the worker node, as it implemented for Fit plugin.

Reply:
Is it possible if we select a socket or sockets during scheduling and the kubelet allocates the pod on the socket. I think after we introduce NUMA into scheduling, the socket may become the minimum scheduling unit.

Reply:
Socket it's another term, several NUMA nodes could be on one socket. Here we're talking about NUMA nodes, yes as I said, if we would manage to add some annotation it's possible for kubelet to make its decision based on scheduler's evaluation.

Reply:
Can we introduce some common approach to control the behaviour on how kubelet set cgroups, including cpuset, cfs and etc.

Reply:
I think external resource management like https://github.com/nokia/CPU-pooler or https://github.com/intel/cri-resource-manager/ could address cgroup management issue.